Contract maintenance

DavidVorick opened the pull request:
The `threadedContractMaintenance` thread was pretty opaque, and had some areas related to funding that were not very understandable, and probably not correct. I refactored the way that the loop checks for funds remaining, and I broke some of the more complex pieces of logic out into their own functions.

I think the tests are passing, though they actually keep timing out on my machine. The siatest renter tests seem to be taking more than 500 seconds, but not because one is deadlocked, rather by the time the later tests are going, the earlier tests have already pushed the timer over 400 seconds.

MSevey commented on modules/renter/contractor/contracts.go:
```diff
@@ -283,6 +359,132 @@ func (c *Contractor) managedRenew(sc *proto.SafeContract, contractFunding types.
+	}
+
+	// Calculate endHeight for renewed contracts
+	endHeight := currentPeriod + allowance.Period
```
there is a function [`contractEndHeight()` ](https://github.com/NebulousLabs/Sia/blob/8da8220708ed647b7fa9e466a0b2ee5355666e47/modules/renter/contractor/contracts.go#L27) which can be called on the `Contractor`.  We should probably use that instead of manually calculating the endHeight in multiple places right?

DavidVorick replied:
contractEndHeight needs a lock

MSevey commented on modules/renter/contractor/contracts.go:
```diff
@@ -331,185 +533,94 @@ func (c *Contractor) threadedContractMaintenance() {
-	// Figure out which contracts need to be renewed, and while we have the
-	// lock, figure out the end height for the new contracts and also the amount
-	// to spend on each contract.
+	// Create the renewSet and refreshSet, each a list of contracts that need to
```
`each is a list` would read more easily

MSevey commented on modules/renter/contractor/contracts.go:
```diff
@@ -331,185 +533,94 @@ func (c *Contractor) threadedContractMaintenance() {
-			fundsAvailable = types.ZeroCurrency
-		}
-	}
+	endHeight := currentPeriod + allowance.Period
```
same here about using `contractEndHeight()`

MSevey commented on modules/renter/contractor/contracts.go:
```diff
@@ -331,185 +533,94 @@ func (c *Contractor) threadedContractMaintenance() {
+	// from the contractor, build those up under a lock so that the rest of the
+	// function can execute without lock contention.
 	c.mu.RLock()
 	currentPeriod := c.currentPeriod
```
shouldn't need to define `currentPeriod` if we use `contractEndHeight()`

DavidVorick replied:
" without lock contention"

Commit:
make estTxnSize of proto package a modules constant

Commit:
refactor threadedContractMaintenance to be more evidently correct

MSevey committed:
Update max upload and max storage prices for contractor to be larger for development and testing to fix failing siatests

Commit:
set a minimum price for contracts that get renewed

Commit:
fix grammar in a comment

Commit:
change hardcoded value into a constant

Previously at least 20% of the per-contract allowance was required to be
spent on a new contract. That amount was hardcoded, now it's a constant
set to 15%

Commit:
Merge branch 'master' into contract-maintenance
